edge/pkg/clusterd/helper/helper.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/kubeedge/kubeedge/edge/pkg/clusterd/util"
+)
+
+func GetLocalClusterScopeResourceNames(resType string, label string) []string {
+	labelOption := ""
+	if len(label) > 0 {
+		labelOption = "-l " + label
+	}
+	getResourceCmd := fmt.Sprintf(" %s get %s -o json %s --kubeconfig=%s | jq -r '.items[] | [.metadata.name] | @tsv' ", config.Config.KubectlCli, resType, labelOption, config.Config.Kubeconfig)
+	output, err := util.ExecCommandLine(getResourceCmd)
+	if err != nil {
+		klog.Errorf("Failed to get %v: %v", resType, err)
+		return []string{err.Error()}
+	}
+
+	names := []string{}
+	for _, o := range strings.Split(output, "\n") {
+		name := strings.TrimSpace(o)
+		if len(name) > 0 {
+			names = append(names, name)
+		}
+	}
+
+	return names
+}
+
+func GetMissionByName(name string) (*edgeclustersv1.Mission, error) {
+	getMissionCmd := fmt.Sprintf("%s get mission %s --kubeconfig=%s -o json ", config.Config.KubectlCli, name, config.Config.Kubeconfig)
+	output, err := util.ExecCommandLine(getMissionCmd)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get mission %v: %v", name, err)
+	}
+
+	var mission edgeclustersv1.Mission
+	err = json.Unmarshal([]byte(output), &mission)
+	if err != nil {
+		return nil, err
+	}
+
+	return &mission, nil
+}
+
+func TestClusterReady() bool {
+	testClusterCommand := fmt.Sprintf("%s cluster-info --kubeconfig=%s", config.Config.KubectlCli, config.Config.Kubeconfig)
+	if _, err := util.ExecCommandLine(testClusterCommand); err != nil {
+		klog.Errorf("The cluster is unreachable: %v", err)
+		return false
+	}
+
+	return true
+}
+
+// this is a hacky way for PoC only
+func AnalyzeMissionContent(content string) (kind string, name string, namespace string) {
+	for _, line := range strings.Split(content, "\n") {
+		words := strings.Split(strings.TrimSpace(line), " ")
+		if strings.Contains(line, "kind:") && kind == "" {
+			kind = words[len(words)-1]
+			continue
+		}
+		if strings.Contains(line, "name:") && name == "" {
+			name = words[len(words)-1]
+			continue
+		}
+		if strings.Contains(line, "namespace:") && namespace == "" {
+			namespace = words[len(words)-1]
+			continue
+		}
+	}
+	return
+}
+
+func EqualClusterReferences(a []edgeclustersv1.GenericClusterReference, b []edgeclustersv1.GenericClusterReference) bool {
+	if a == nil && b == nil {
+		return true
+	}
+	if a == nil && len(b) == 0 {
+		return true
+	}
+	if b == nil && len(a) == 0 {
+		return true
+	}
+
+	// for other cases, use the regular array compare
+	return reflect.DeepEqual(a, b)
+}
+
+func EqualMaps(a map[string]string, b map[string]string) bool {
+	if a == nil && b == nil {
+		return true
+	}
+	if a == nil && len(b) == 0 {
+		return true
+	}
+	if b == nil && len(a) == 0 {
+		return true
+	}
+
+	// for other cases, use the regular map compare
+	return reflect.DeepEqual(a, b)
+}
+
+// for some reason we still need to find out, the same mission spec objects may be no longer deep-equal.
```

Comment:
please create an issue for this. very intriguing to find out.

Reply:
Now I know this issue is due to Json encoder. Json encoding a nil array into an empty array, thus the deep-equal conditions no long hold. 

I rewrote the comment to explain this.